homeassistant/components/flux_led/light.py

```diff
@@ -308,101 +376,103 @@ def effect(self):
+
+            self._bulb.setRgbw(
+                r=color_red,
+                g=color_green,
+                b=color_blue,
             )
 
-        # Random color effect
-        if effect == EFFECT_RANDOM:
-            self._bulb.setRgb(
-                random.randint(0, 255), random.randint(0, 255), random.randint(0, 255)
+            self._hs_color = color_util.color_RGB_to_hs(
+                color_red,
+                color_green,
+                color_blue,
             )
-            return
+            self._last_update = time.time()
 
-        if effect == EFFECT_CUSTOM:
-            if self._custom_effect:
-                self._bulb.setCustomPattern(
-                    self._custom_effect[CONF_COLORS],
-                    self._custom_effect[CONF_SPEED_PCT],
-                    self._custom_effect[CONF_TRANSITION],
-                )
             return
 
-        # Effect selection
         if effect in EFFECT_MAP:
             self._bulb.setPresetPattern(EFFECT_MAP[effect], self._effect_speed)
-            return
 
-        # Preserve current brightness on color/white level change
-        if brightness is None:
-            brightness = self.brightness
+        if not brightness:
+            brightness = self._last_brightness
+
+        self._last_brightness = brightness
+        self._brightness = brightness
+
+        if not rgb:
+            rgb = color_util.color_hs_to_RGB(*self._last_hs_color)
 
-        # Preserve color on brightness/white level change
-        if rgb is None:
-            rgb = self._bulb.getRgb()
+        self._hs_color = color_util.color_RGB_to_hs(*tuple(rgb))
 
-        if white is None and self._mode == MODE_RGBW:
+        if not white and self._mode == MODE_RGBW:
             white = self.white_value
 
-        # handle W only mode (use brightness instead of white value)
         if self._mode == MODE_WHITE:
             self._bulb.setRgbw(0, 0, 0, w=brightness)
 
-        # handle RGBW mode
         elif self._mode == MODE_RGBW:
             self._bulb.setRgbw(*tuple(rgb), w=white, brightness=brightness)
 
-        # handle RGB mode
         else:
             self._bulb.setRgb(*tuple(rgb), brightness=brightness)
 
-    def turn_off(self, **kwargs):
-        """Turn the specified or all lights off."""
-        self._bulb.turnOff()
-
-    def update(self):
-        """Synchronize state with bulb."""
-        if not self.available:
-            try:
-                self._connect()
-                self._error_reported = False
-            except OSError:
-                self._disconnect()
-                if not self._error_reported:
-                    _LOGGER.warning(
-                        "Failed to connect to bulb %s, %s", self._ipaddr, self._name
-                    )
-                    self._error_reported = True
-                return
+        self._state = True
+        self._last_update = time.time()
+
+    async def async_turn_off(self, **kwargs):
```

Comment:
Why are we changing the light turn on/off methods? We shouldn't do that in this PR.

Reply:
Why not? This is the refactoring of the code which corrects the underlying issues with some flux_led lights.

Reply:
This PR should be about the coordinator which fetches state. The actuator methods send state. These are different things. What's the relation?

Reply:
Overall the PR (including all parts) does four things... add the config_flow (done in previous merge), add data update coordinator and refactor the code to fix issues (this PR), and add the custom service to replace the old custom_effect functionality which goes away with the config_flow change.

I can break the turn_on/turn_off out into another PR but it will need to be grouped together with all the rest for this to be complete. Let me know your preference - I thought that change was small enough to include in this PR.

Reply:
Please explain the changes here. It seems we're changing how the light properties are computed and not just switching to use the coordinator.

We shouldn't mix different things in the dame PR unless required. Preferably we should only introduce the coordinator and not change how the light properties are handled.

Reply:
The primary changes in the turn_on and turn_off are that we now store the last state in the turn_off method rather than when turning on (brightness, color) to capture changes which were made outside of Home Assistant (ie. with a remote control) and the removal of Color Temperature settings which do not work on the newer lights.

Additionally I added the delay with self._last_update to prevent the switch from flipping back and forth - I can change that to assumed state but I need an example to look at on that one.

I'm not sure why Git has highlighted such significant changes here, when I scroll through the file changes it is significantly the same.

Reply:
The old turn off method just made one call to turn off the light. We shouldn't change this behavior in the same PR as we're introducing the coordinator.

Reply:
OK - removed the changed turn_off and went back to just `turnOff()` call. I did keep the storing of the last states however and included the state update and `async_write_ha_state` to avoid "flipping" behaviour in the UI.